### PR TITLE
Docker build for FORCE v3.0

### PR DESCRIPTION
**Pull request by fegyi001**
First of all, I'm so happy that FORCE has a new major version, and also that it is on Github now.
I added support for Docker builds which hopefully makes possible for easier usage of this awesome software.

### CODE REVIEW

File: docs/source/setup/depend.rst [hunk 1]
```diff
@@ -110,8 +110,10 @@ FORCE can also be installed on other Liunx distributions (e.g. CentOS). The inst
      cd ~/src/opencv
      wget https://github.com/opencv/opencv/archive/4.1.0.zip
      unzip 4.1.0.zip
+     cd opencv-4.1.0
      mkdir build
      cd build
      cmake -D CMAKE_BUILD_TYPE=Release -D CMAKE_INSTALL_PREFIX=/usr/local ..
      make -j
      sudo make install
+     sudo make clean
```
Comment (davidfrantz) [inline docs/source/setup/depend.rst]: Is sudo really necessary here?

**Comment by davidfrantz** [comment]
Dear Gergely,
thanks for your commits! This might be a nice improvement!
I haven't used Docker before, thus I am not entirely sure how the procedure of installing works. However, I quickly went over your commit and I think there is a "chicken/egg" issue with the suggested installation if the docker part merges into the same repository. 
Probably, we can find a way to make this more self-contained, and e.g. get rid of the duplicated files with minor changes. Once I have some time, I will make some more comments on this. 
Cheers,
David

File: docs/source/setup/depend.rst (hunk 1 above)
Comment (fegyi001) [reply docs/source/setup/depend.rst]: A clean after installing OpenCV is highly desirable in my opinion. Without cleaning after install, the Docker image was beyond 6 GB (without it around 1.8 GB ). Here I only added for helping the documentation. Sudo I guess is needed if it was needed for the make install command.

Comment (davidfrantz) [reply docs/source/setup/depend.rst]: Agreed, cleaning makes sense. But sudo wasn't used for the build, only for the install ;) Cleaning the temporary files in the build directory should work without sudo.

**Comment by fegyi001** [comment]
Docker is good if you don't want to bother with the installation ("copy this, run that, alter this, install from here" etc. etc). You will became OS independent, the docker container should work exactly the same way without any side effects no matter if you run it on Windows, Linux or MacOS. If someone uses a docker image from Docker hub, all she/he has to do is type one single line of command into a terminal, the complete image will be downloaded locally, and whoa, it will just work. And you save a lot of time: you only need to download the image (1-2 mins) instead of installing it (unpredictable how long it takes).

Docker is really really widely used, so imho if a Dockerized FORCE is available many would use it instead of installing every dependencies in the right order (with trying to fix any errors and start over and over again). As a software engineer I almost always use the Docker version of a software if there is such.

**Comment by fegyi001** [comment]
I don't really understand what you mean by the "chicken/egg" analogue. Maybe that in the docker folder there is also a Makefile and a const-cl.h file? If that is what bothers you I can fix that: I can move the Dockerfile to the root folder, and in that case it will be able to see src/ during a Docker build. No duplicates that way.

**Comment by davidfrantz** [comment]
OK Gergely, as said, I never used Docker, but it sounds very nice. So let's try to incorporate this.

Before starting, just allow me a small question out of curiosity and not related to the following: are dockerized programs as fast as locally built programs?

There are two things that bother me a bit with the provided commit:
1) as you pointed out, you have copies of const-cl.h and Makefile. And as far as I can see, you only changed the lines that deal with the SPLITS installation. I'd like to avoid such redundancy in the repository as future code maintenance will become more complicated and error-prone (any change needs to be made in two files). However, if the docker commit will be integrated, there shouldn't be a need to have static copies. The changes to const-cl.h and Makefile can be made with bash commands (note that I have just updated the installation docs with a way to do this more automatically).
2) The chicken/egg problem I mentioned refers to your docker Makefile (l. 40). In there, you clone the v. 3 FORCE repository. However, if the docker commit is integrated, users will already have the code (probably even a newer version, there already is v. 3.0.1 online). Instead, the v. 3.0 version will be cloned and installed. 

If you can take care of these two points, I'd be happy to pull your commit.

Cheers,
David

File: src/cross-level/const-cl.h [hunk 2]
```diff
@@ -233,7 +233,7 @@ enum { _CLOCK_NULL_, _CLOCK_TICK_, _CLOCK_TOCK_, _CLOCK_LENGTH_ };
 //#define FORCE_DEBUG
 //#define FORCE_DEV
 
-//#define SPLITS
+#define SPLITS
```
Comment (fegyi001) [inline src/cross-level/const-cl.h]: I think it would be nice if SPLITS was enabled by default.

File: Makefile [hunk 3]
```diff
@@ -29,15 +29,15 @@ BINDIR=/usr/local/bin
 # Libraries
 GDAL=-I/usr/include/gdal -L/usr/lib -Wl,-rpath=/usr/lib
 GSL=-I/usr/include/gsl -L/usr/lib/x86_64-linux-gnu -Wl,-rpath=/usr/lib/x86_64-linux-gnu -DHAVE_INLINE=1 -DGSL_RANGE_CHECK=0
-CURL=-I/usr/include/curl -L/usr/lib/x86_64-linux-gnu -Wl,-rpath=/usr/lib/x86_64-linux-gnu
-#CURL=-I/usr/include/x86_64-linux-gnu/curl -L/usr/lib/x86_64-linux-gnu -Wl,-rpath=/usr/lib/x86_64-linux-gnu
+# CURL=-I/usr/include/curl -L/usr/lib/x86_64-linux-gnu -Wl,-rpath=/usr/lib/x86_64-linux-gnu
+CURL=-I/usr/include/x86_64-linux-gnu/curl -L/usr/lib/x86_64-linux-gnu -Wl,-rpath=/usr/lib/x86_64-linux-gnu
 OPENCV=-I/usr/local/include/opencv4 -L/usr/local/lib -Wl,-rpath=/usr/local/lib
-#SPLITS=-I/usr/local/include/splits -L/usr/local/lib -Wl,-rpath=/usr/local/lib
+SPLITS=-I/usr/local/include/splits -L/usr/local/lib -Wl,-rpath=/usr/local/lib
```
Comment (fegyi001) [inline Makefile]: I think it would be nice if SPLITS was enabled by default.

Comment (fegyi001) [inline Makefile]: I think it would be nice if SPLITS was enabled by default.

File: Makefile [hunk 4]
```diff
@@ -29,15 +29,15 @@ BINDIR=/usr/local/bin
 # Libraries
 GDAL=-I/usr/include/gdal -L/usr/lib -Wl,-rpath=/usr/lib
 GSL=-I/usr/include/gsl -L/usr/lib/x86_64-linux-gnu -Wl,-rpath=/usr/lib/x86_64-linux-gnu -DHAVE_INLINE=1 -DGSL_RANGE_CHECK=0
-CURL=-I/usr/include/curl -L/usr/lib/x86_64-linux-gnu -Wl,-rpath=/usr/lib/x86_64-linux-gnu
-#CURL=-I/usr/include/x86_64-linux-gnu/curl -L/usr/lib/x86_64-linux-gnu -Wl,-rpath=/usr/lib/x86_64-linux-gnu
+# CURL=-I/usr/include/curl -L/usr/lib/x86_64-linux-gnu -Wl,-rpath=/usr/lib/x86_64-linux-gnu
+CURL=-I/usr/include/x86_64-linux-gnu/curl -L/usr/lib/x86_64-linux-gnu -Wl,-rpath=/usr/lib/x86_64-linux-gnu
```
Comment (fegyi001) [inline Makefile]: In my case only this line worked for curl with the official Ubuntu 18.04 LTS version

**Comment by fegyi001** [comment]
I did what you asked for. The docker folder will only contain a more detailed Readme. Now the Docker image will build FORCE from the src/ folder. Maybe it could be used for testing for you as well? I would gladly help you with that.

**Comment by fegyi001** [comment]
I made back SPLITS opt-out by default

File: Makefile (hunk 4 above)
Comment (davidfrantz) [reply Makefile]: Yes, the other line works for 16.04 LTS. I just commited a change to master that works with both.

File: docs/source/setup/install.rst [hunk 5]
```diff
@@ -96,5 +96,30 @@ Installation with optional software
...
+
+* Use prebuilt image
+
+  The easiest way to use FORCE with Docker is to use a prebuilt image pulled from `Docker hub <https://hub.docker.com/>`_ with the following command:
   
-  
\ No newline at end of file
+  ``docker pull fegyi001/force:latest``
+
+  This downloads a fully featured FORCE v3.0 on your local machine including SPLITS.
```
Comment (davidfrantz) [inline docs/source/setup/install.rst]: Is this latest or 3.0?

**Comment by davidfrantz** [comment]
Regarding SPLITS: I thought long about making it default. But there are some problems involved. Some people have experienced difficulties in running it (some library conflicts within Armadillo). And the developer has left academia and no longer maintains the software, and doesn't answer emails. Therefore, I think I will keep it optional.

**Comment by davidfrantz** [comment]
Please note that I have created a small script that can enable/disable SPLITS (see installation). You could also add this to your docker installation script.

File: docs/source/setup/install.rst (hunk 5 above)
Comment (fegyi001) [reply docs/source/setup/install.rst]: I fixed that

**Comment by fegyi001** [comment]
> Please note that I have created a small script that can enable/disable SPLITS (see installation). You could also add this to your docker installation script.

I looked up the documentation, which says there is a src/force/splits.sh file, but I could not find it anywhere.

**Comment by davidfrantz** [comment]
Thanks for your contribution, Gergely! And also for the changes, much appreciated.
Best,
David

**Pull request merged**